compactcert/builder.go

```diff
@@ -0,0 +1,332 @@
+	}
+
+	if builder.Present(pos) {
+		// Signature already part of the builder, ignore.
+		return network.Ignore, nil
+	}
+
+	err := builder.Add(pos, sfa.Sig, true)
+	if err != nil {
+		return network.Disconnect, err
+	}
+
+	err = ccw.db.Atomic(func(ctx context.Context, tx *sql.Tx) error {
+		return addPendingSig(tx, sfa.Round, pendingSig{
+			signer:       sfa.Signer,
+			sig:          sfa.Sig,
+			fromThisNode: sender == nil,
+		})
+	})
+	if err != nil {
+		return network.Ignore, err
+	}
+
+	return network.Broadcast, nil
+}
+
+func (ccw *Worker) builder() {
+	// We clock the building of compact certificates based on new
+	// blocks.  This is because the acceptable compact certificate
+	// size grows over time, so that we aim to construct an extremely
+	// compact certificate upfront, but if that doesn't work out, we
+	// will settle for a larger certificate.  New blocks also tell us
+	// if a compact cert has been committed, so that we can stop trying
+	// to build it.
+	for {
+		nextrnd := ccw.ledger.Latest() + 1
+		select {
+		case <-ccw.ctx.Done():
+			return
+
+		case <-ccw.ledger.Wait(nextrnd):
+			// Continue on
+		}
+
+		hdr, err := ccw.ledger.BlockHdr(nextrnd)
+		if err != nil {
+			ccw.log.Warnf("ccw.builder: BlockHdr(%d): %v", nextrnd, err)
+		} else {
+			ccw.deleteOldSigs(hdr.CompactCertLastRound)
+		}
+
+		// Broadcast signatures based on the previous block that
+		// was agreed upon.  This ensures that, if we send a signature
+		// for block R, nodes will have already verified block R, because
+		// block R+1 has been formed.
+		proto := config.Consensus[hdr.CurrentProtocol]
+		ccw.broadcastSigs(nextrnd-1, proto)
+
+		ccw.tryBuilding()
+	}
+}
+
+// broadcastSigs periodically broadcasts pending signatures for rounds
+// that have not been able to form a compact certificate.
+//
+// Signature re-broadcasting happens in periods of proto.CompactCertRounds
+// rounds.
+//
+// In the first half of each such period, signers of a block broadcast their
+// own signatures; this is the expected common path.
+//
+// In the second half of each such period, any signatures seen by this node
+// are broadcast.
+//
+// The broadcast schedule is randomized by the address of the block signer,
+// for load-balancing over time.
+func (ccw *Worker) broadcastSigs(brnd basics.Round, proto config.ConsensusParams) {
+	ccw.mu.Lock()
+	defer ccw.mu.Unlock()
+
+	var roundSigs map[basics.Round][]pendingSig
+	err := ccw.db.Atomic(func(ctx context.Context, tx *sql.Tx) (err error) {
+		if brnd%basics.Round(proto.CompactCertRounds) < basics.Round(proto.CompactCertRounds/2) {
+			roundSigs, err = getPendingSigsFromThisNode(tx)
+		} else {
+			roundSigs, err = getPendingSigs(tx)
+		}
+		return
+	})
+	if err != nil {
+		ccw.log.Warnf("broadcastSigs: getPendingSigs: %v", err)
+		return
+	}
+
+	for rnd, sigs := range roundSigs {
+		for _, sig := range sigs {
+			// Randomize which sigs get broadcast over time.
```

Comment:
How long does it take in expectation to get the whole cert in the common case?

Reply:
I would expect it to take slightly more than `CompactCertRounds/2` (i.e., 64) rounds.  The "slightly more" is to allow one of the relay nodes to broadcast the compact cert transaction, and then to allow that transaction to make it into a block, which probably takes another 1-2 rounds.

One degenerate case might be that we have fewer than 64 participation keys, in which case we might get 100% of the signatures earlier than 64 rounds, so then we'd be able to form the certificate at that point.

If some nodes are offline (and thus we have fewer than 100% sigs even after 64 rounds), the cert will be formed slightly later (somewhere between 64 and 128 rounds), depending on how much less than 100% we get.